<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Contributing" href="contributing.html" /><link rel="prev" title="Installation" href="installation.html" />

    <!-- Generated with Sphinx 7.0.1 and Furo 2023.05.20 -->
        <title>Usage - SEMPER Recode 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=e6660623a769aa55fea372102b9bf3151b292993" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">SEMPER Recode 0.1.0 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">SEMPER Recode 0.1.0 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="readme.html">Semper Recode</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this heading">#</a></h1>
<p>import os
import re
import pickle
import warnings
import pandas as pd
from Bio import SeqIO
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord</p>
<p>‘’’
Global variables declaration</p>
<p>Loads data required for the analysis. Reads various data files and initializes necessary attributes.</p>
<section id="raises">
<h2>Raises<a class="headerlink" href="#raises" title="Permalink to this heading">#</a></h2>
<dl class="simple">
<dt>FileNotFoundError</dt><dd><p>If any of the required data files are not found.</p>
</dd>
</dl>
<p>‘’’
START_CODON = [‘ATG’]
PATH = “data/”</p>
<p># Master dataframe
MASTER_DF_PATH = os.path.join(PATH, ‘master_df_os_2023.csv’)
if not os.path.exists(MASTER_DF_PATH):</p>
<blockquote>
<div><p>raise FileNotFoundError(f”Master dataframe file not found: {MASTER_DF_PATH}”)</p>
</div></blockquote>
<p>MASTER_DF = pd.read_csv(PATH + ‘master_df_os_2023.csv’)</p>
<p># Load codon list
CODON_LIST_PATH = os.path.join(PATH, ‘codon_list.pkl’)
if not os.path.exists(CODON_LIST_PATH):</p>
<blockquote>
<div><p>raise FileNotFoundError(f”Codon list (.pkl) file not found: {CODON_LIST_PATH}”)</p>
</div></blockquote>
<dl class="simple">
<dt>with open(CODON_LIST_PATH, ‘rb’) as file:</dt><dd><p>CODON_LIST = pickle.load(file)</p>
</dd>
<dt>if CODON_LIST == {}:</dt><dd><p>raise ValueError(“codon_list is empty”)</p>
</dd>
</dl>
<p># Load codon dictionary
CODON_DICT_PATH = os.path.join(PATH, ‘codon_dict.pkl’)</p>
<dl class="simple">
<dt>with open(CODON_DICT_PATH, ‘rb’) as file:</dt><dd><p>CODON_DICT = pickle.load(file)</p>
</dd>
</dl>
<p># Four letters codes
FOUR_LETTERS_CODE = list(MASTER_DF[‘4-letters’].unique())  # Get unique four-letter codes from master_df</p>
<p>‘’’
Functions:</p>
<blockquote>
<div><p>&gt; data_prep(self, path)
&gt; process_sequence(self)</p>
<blockquote>
<div><dl class="simple">
<dt>&gt; modify_TIS_in_frame(self, sequence)</dt><dd><p>&gt; find_in_frame(self, sequence)
&gt; efficiency_level(self, sequence)</p>
</dd>
<dt>&gt; modify_TIS_out_of_frame(self, sequence)</dt><dd><p>&gt; find_out_of_frame(self, sequence)
&gt; find_key(self, codon)
&gt; get_aa_key(self, original_codon)</p>
</dd>
</dl>
</div></blockquote>
<p>&gt; to_fasta(self, sequence, output_file_name)
&gt; filtered_sequence_eff(self, efficiency)
&gt; find_lower_eff_sequence(self, efficiency, seq)</p>
</div></blockquote>
<p>‘’’</p>
<dl>
<dt>class SemperRecode:</dt><dd><dl>
<dt>def __init__(self, user_seq):</dt><dd><p>“””
Initializes the SemperRecode object.</p>
<p>user_seq : str</p>
<dl class="simple">
<dt>ValueError</dt><dd><ol class="arabic simple">
<li><p>If input sequence is blank.</p></li>
<li><p>If input sequence contains “u” or “U”</p></li>
</ol>
</dd>
</dl>
<p>“””
self.master_df = MASTER_DF
self.start_codon = START_CODON</p>
<p># Convert user input sequence into str, remove blank space, and capitalize the string
self.seq = str(user_seq).replace(” “, “”).upper()</p>
<p># Check if user input sequence is empty
if len(self.seq) == 0 or self.seq.isspace():</p>
<blockquote>
<div><p>raise ValueError(“No sequence input”)</p>
</div></blockquote>
<p># Raise ValueError if user input sequence contains “u” or “U”
if “U” in self.seq or “u” in self.seq:</p>
<blockquote>
<div><p>raise ValueError(f”U or u found in the input sequence {self.seq}”)</p>
</div></blockquote>
<p># Raise ValueError if the sequence contains any character other than A, T, C, G
pattern = r’[^ATCG]’
if re.search(pattern, self.seq):</p>
<blockquote>
<div><p>raise ValueError(“Invalid character found in the string.”)</p>
</div></blockquote>
</dd>
<dt>def process_sequence(self):</dt><dd><p>“””
Uses the nucleotide sequence input through the constructor and returns the modified sequence with lower efficiency (if any).
Loop through modify_TIS_in_frame() and modify_TIS_out_of_frame() to ensure there’s no out-of-frame internal AUG and that the
expressional level is the lowest it could possibly be</p>
<p>None</p>
<dl class="simple">
<dt>replace_sequence<span class="classifier">str</span></dt><dd><p>String of modified sequences of the input sequence.</p>
</dd>
</dl>
<p>None</p>
<p>“””
replace_sequence = self.seq
round = 0</p>
<p># Loop through modify_TIS_in_frame() and modify_TIS_out_of_frame() while there is out-of-frame AUGs or the loop hasn’t reached 10 times
while self.find_out_of_frame(replace_sequence) != [] and round &lt; 10:</p>
<blockquote>
<div><p># Find modified sequence which is returned by modify_TIS_in_frame()
temp = self.modify_TIS_in_frame(self.seq)
replace_sequence = self.modify_TIS_out_of_frame(temp)
round += 1</p>
</div></blockquote>
<p>return replace_sequence</p>
</dd>
<dt>def modify_TIS_in_frame(self, sequence):</dt><dd><p>‘’’
Takes in sequence (str) and return modified sequence with lower TIS efficiency (str)
by getting all the index location of internal AUG from find_in_frame() and modify the sequence accordingly</p>
<blockquote>
<div><p>sequence : str</p>
</div></blockquote>
<blockquote>
<div><p>modified_sequence (str)</p>
</div></blockquote>
<p>‘’’
new_seq = list(sequence)
index = self.find_in_frame(sequence) # Get the indices of in-frame AUG(s)
df = self.master_df</p>
<p>‘’’
Convert the sequence to a list (new_seq) so that’s it’s mutable and when we iterate through the sequence,
we can replace the TIS sequence with the modified sequence with lower efficiency according to the index
of AUG return by find_in_frame()</p>
<p>This way, we’ll iterate through the updated list of sequence (new_seq) with modified sequence
‘’’
for pos in index:</p>
<blockquote>
<div><p># Ignore the first and last AUG
if pos != 0 and pos != len(sequence) - 3:</p>
<blockquote>
<div><dl class="simple">
<dt>if pos == len(sequence) - 3:</dt><dd><p>warnings.warn(“There’s an AUG at the end of the sequence which cannot be modified”)</p>
</dd>
</dl>
<p>sub_1 = ‘’.join(new_seq[pos-6:pos+5])
internal_TIS_seq = Seq(sub_1)</p>
<p>sub_2 = ‘’.join(new_seq[pos-6:pos+6])
aa4 = Seq(sub_2).translate()</p>
<p># Get the current efficiency level
current_eff = self.efficiency_level(internal_TIS_seq)</p>
<p>filtered = df[df[‘4-letters’] == str(aa4)]
new_eff = filtered[‘efficiency’].iloc[0]</p>
<p>‘’’
Replace the designated TIS sequence with the modified sequence
If a new sequence with a lower efficiency is not found,
print a message telling the user to consider mutate/remove the sequence
‘’’
new_seq[pos-6:pos+6] = filtered[“4-codons”].iloc[0]</p>
<dl class="simple">
<dt>if(int(new_eff) == current_eff):</dt><dd><p>warnings.warn(f”No sequence with lower efficiency is found for {internal_TIS_seq}, consider mutate/remove the sequence”)</p>
</dd>
</dl>
</div></blockquote>
</div></blockquote>
<p>return ‘’.join(new_seq)</p>
</dd>
<dt>def find_in_frame(self, sequence):</dt><dd><p>“””
Takes in a sequence (str) and returns a list of indices where the AUG codon is found.</p>
<dl class="simple">
<dt>sequence<span class="classifier">str</span></dt><dd><p>The input sequence to search for AUG codons.</p>
</dd>
</dl>
<dl class="simple">
<dt>pos<span class="classifier">list</span></dt><dd><p>A list of integers representing the indices of in-frame AUG codons in the sequence.</p>
</dd>
</dl>
<p>“””
pos = []  # position(s) of in-frame AUG in the string (sequence)</p>
<p># Loop through codon by codon (start from 0, end at last codon, increase by 3)
for i in range(0, len(sequence)-2, 3):</p>
<blockquote>
<div><p>codon = sequence[i:i+3]</p>
<dl class="simple">
<dt>if codon in self.start_codon:</dt><dd><p>pos.append(i)</p>
</dd>
</dl>
</div></blockquote>
<p>return pos</p>
</dd>
<dt>def efficiency_level(self, sequence):</dt><dd><p>‘’’
Takes in sequence and return efficiency level</p>
<dl class="simple">
<dt>sequence<span class="classifier">str</span></dt><dd><p>The input sequence to find efficiency level.</p>
</dd>
</dl>
<dl class="simple">
<dt>eff<span class="classifier">int</span></dt><dd><p>An interger represent the efficiency level of the sequence when
being compared with the master dataframe</p>
</dd>
</dl>
<p>‘’’
df = self.master_df</p>
<p>‘’’
Find the matching row in master_df with the same value in tis-sequence
as the input sequence
‘’’
match = df[df[“tis-sequence”] == str(sequence)]</p>
<dl class="simple">
<dt>if match.empty:</dt><dd><p># No matching is found
raise ValueError(f”The sequence {sequence} is not found in the dataframe”)</p>
</dd>
</dl>
<p>return match[‘efficiency’].iloc[0]</p>
</dd>
<dt>def modify_TIS_out_of_frame(self, sequence):</dt><dd><p>‘’’
Takes in sequence (str) and return modified sequence
with lower TIS efficiency (str)</p>
<blockquote>
<div><p>sequence : str</p>
</div></blockquote>
<blockquote>
<div><p>modified_sequence (str)</p>
</div></blockquote>
<p>‘’’</p>
<p>new_seq = list(sequence)
index = self.find_out_of_frame(sequence) # Get the indices of out-of-frame AUG(s)</p>
<p>‘’’
Iterate through sequene and modify sequence to get rid of any out-of-frame AUG(s)
using the index from find_out_of_frame()
‘’’</p>
<dl>
<dt>for pos in index:</dt><dd><p># Break the it down into 2 codons according to its proper codon frame
start = pos - pos%3
first_aa = sequence[start : start + 3]
second_aa = sequence[start + 3 : start + 6]</p>
<p>‘’’
Compare which codon (the old and the new one) has the least difference in fraction value
When the pair is found, replace the old codon with the new codon
‘’’
# Get the fraction value of the first old codon
first_aa_key = self.return_key(first_aa)
first_old_val = CODON_DICT[first_aa_key][first_aa]</p>
<p># Get the fraction value of the second old codon
second_aa_key = self.return_key(second_aa)
second_old_val = CODON_DICT[second_aa_key][second_aa]</p>
<p># Get the codons and fraction values of the two new codons
first_new_aa, first_new_val = self.get_aa_alternative(first_aa)
second_new_aa, second_new_val = self.get_aa_alternative(second_aa)</p>
<p># Find the difference between each pair of codons (old and new)
first_aa_diff = abs(first_new_val - first_old_val)
second_aa_diff = abs(second_new_val - second_old_val)</p>
<p>‘’’
Replace the codon with the least fraction difference
i.e. if the fraction difference between the old and the new amino acids of first aa is greater,
replace the second codon with the new codon
‘’’</p>
<dl class="simple">
<dt>if second_aa_diff &gt; first_aa_diff:</dt><dd><p>new_seq[start : start + 3] = first_new_aa # Replace the first codon</p>
</dd>
<dt>else:</dt><dd><p>new_seq[start + 3 : start + 6] = second_new_aa # Replace the second codon</p>
</dd>
</dl>
</dd>
</dl>
<p>return ‘’.join(new_seq)</p>
</dd>
<dt>def return_key(self, codon):</dt><dd><p>‘’’
Takes in codon sequence and return the key (str) which is the abbreviation of the codon</p>
<blockquote>
<div><p>codon : str</p>
</div></blockquote>
<blockquote>
<div><p>key (str)</p>
</div></blockquote>
<p>‘’’
key = “”</p>
<dl class="simple">
<dt>for char, value in CODON_LIST.items():</dt><dd><dl class="simple">
<dt>if codon in value:</dt><dd><p>key = char
break</p>
</dd>
</dl>
</dd>
</dl>
<p>return key</p>
</dd>
<dt>def find_out_of_frame(self, sequence):</dt><dd><p>“””
Takes in a sequence (str) and returns a list of indices where the out-of-frame AUG codon is found.</p>
<dl class="simple">
<dt>sequence<span class="classifier">str</span></dt><dd><p>The input sequence to search for AUG codons.</p>
</dd>
</dl>
<dl class="simple">
<dt>pos<span class="classifier">list</span></dt><dd><p>A list of integers representing the indices of out-of-frame AUG codons in the sequence.</p>
</dd>
</dl>
<p>“””
pos = []
index = -1</p>
<dl>
<dt>for codon in self.start_codon:</dt><dd><p>index = -1</p>
<dl>
<dt>while True:</dt><dd><p>index = sequence.find(codon, index + 1)
if index == -1:</p>
<blockquote>
<div><p>break</p>
</div></blockquote>
<dl class="simple">
<dt>if index%3 != 0:</dt><dd><p>pos.append(index)</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
<p>return pos</p>
</dd>
<dt>def get_aa_alternative(self, original_codon):</dt><dd><p>‘’’
Takes in original_codon (ex: ‘GCC’, ‘GAG’, ‘TCA’) then return:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>the codon which produce the key amino acid with highest fraction (other than the original codon itself)</p></li>
<li><p>The fraction of that codon</p></li>
</ol>
</div></blockquote>
<p>Sample code: get_aa_key(‘GCC’)
Output: ‘GCT’, 0.2301414614526459</p>
<blockquote>
<div><p>original_codon : str</p>
</div></blockquote>
<blockquote>
<div><p>codon (str)
value (int)</p>
</div></blockquote>
<p>‘’’
# Find the corresponding amino acid (ex: ‘A’, ‘R’, ‘N’)
aa = self.return_key(original_codon)</p>
<p># Get all the possible codons which product the key (ex: [‘CAC’: 0.5742015169781323, ‘CAT’: 0.4257984830218677])
codons = list(CODON_DICT[aa].keys())
index = 0</p>
<p># If the codon is the same as the original one and there’s an alternative, get to the next one
if codons[0] == original_codon and len(codons) &gt; 1:</p>
<blockquote>
<div><p>index += 1</p>
</div></blockquote>
<p># The fraction of the new codon
value = CODON_DICT[aa][codons[index]]</p>
<p>return codons[index], value</p>
</dd>
<dt>def to_fasta(self, sequence, output_file_name):</dt><dd><p>‘’’
Takes in a list of modified sequences and converts them back to FASTA format for exporting to users.</p>
<dl class="simple">
<dt>sequence<span class="classifier">list</span></dt><dd><p>The modified list of sequences to be converted to a FASTA file.</p>
</dd>
<dt>output_file_name<span class="classifier">str</span></dt><dd><p>The desired name of the output FASTA file.</p>
</dd>
</dl>
<p>None</p>
<dl class="simple">
<dt>ValueError</dt><dd><p>If the sequence is empty.</p>
</dd>
</dl>
<p>‘’’
if not sequence:</p>
<blockquote>
<div><p>raise ValueError(“Sequence list is empty, unable to export”)</p>
</div></blockquote>
<p>output = []</p>
<dl class="simple">
<dt>for i, seq in enumerate(sequence):</dt><dd><p>temp = SeqRecord(Seq(seq), id=f’Seq{i+1}’, description=’’)
output.append(temp)</p>
</dd>
</dl>
<p>‘’’
i = index as wel iterated through sequence</p>
<p>In each iteration, a new SeqRecord object (Seq object but with additional metadata)
is being created</p>
<dl class="simple">
<dt>Metadata:</dt><dd><p>id: The sequence original sequence name stored in self.seq_id
description: None</p>
</dd>
</dl>
</dd>
</dl>
</dd>
<dt>1</dt><dd><blockquote>
<div><p>‘’’
output_file = “tests/sample_file/” + output_file_name + “.fasta”</p>
<dl class="simple">
<dt>with open(output_file, ‘w’) as file:</dt><dd><p>SeqIO.write(sequence, file, ‘fasta’)</p>
</dd>
</dl>
<p>‘’’
Output:
&gt;Sequence1
ATCGATCGATCG</p>
<p>&gt;Sequence2
GCTAGCTAGCTA</p>
<p>&gt;Sequence3
TGCAATGCAATG
‘’’</p>
</div></blockquote>
<blockquote>
<div><p>Exploratory analysis</p>
</div></blockquote>
<p>‘’’
def filtered_sequence_eff(self, efficiency):</p>
<blockquote>
<div><p>“””
Takes in efficiency level input and finds the sequences in the dataframe by filtering,
then displays all the possible sequences that have an efficiency lower than the given value
and the percentage of found and missing amino acid sequences in the form of a dataframe.</p>
<dl class="simple">
<dt>efficiency<span class="classifier">float</span></dt><dd><p>The efficiency level to filter the dataframe.</p>
</dd>
</dl>
<dl class="simple">
<dt>pd.DataFrame</dt><dd><p>A dataframe containing the filtered sequences.</p>
</dd>
</dl>
<p>“””
df = self.master_df</p>
<p>filtered = df[df[‘efficiency’] &lt; efficiency]
count = filtered.shape[0]
percent = (count / df.shape[0]) * 100</p>
<p>groups = filtered.groupby(‘4-codons’)
seq_names = groups[‘4-letters’].apply(lambda x: ‘, ‘.join(x.astype(str).unique())).tolist()</p>
<p>set1 = set(self.four_letter_codes)
set2 = set(seq_names)</p>
<p>missing_elements = list(set1 - set2)</p>
<p>found_percent = len(set2) / len(set1) * 100</p>
<p>print(f”Total: {count} sequence(s) ({percent:.2f}%)”)
print(f”Found ({found_percent:.2f}%): {’, ‘.join(set2)}”)
print(f”Missing ({(100-found_percent):.2f}%): {’, ‘.join(missing_elements)}n”)</p>
<dl class="simple">
<dt>for each in groups:</dt><dd><p>print(each, “n”)</p>
</dd>
</dl>
<p>return filtered.DataFrame</p>
</div></blockquote>
<dl>
<dt>def find_lower_eff_sequence(self, efficiency, seq):</dt><dd><p>“””
Takes in a nucleotide sequence and checks if there is another sequence that produces
the same protein but with a lower efficiency level than the input level.</p>
<dl class="simple">
<dt>efficiency<span class="classifier">float</span></dt><dd><p>The efficiency level to filter the dataframe.</p>
</dd>
<dt>seq<span class="classifier">Bio.Seq</span></dt><dd><p>nucleotide sequence</p>
</dd>
</dl>
<dl class="simple">
<dt>ValueError</dt><dd><p>If the given sequence doesn’t have replacing sequence with efficiency lower
than the input efficiency level.</p>
</dd>
</dl>
<dl class="simple">
<dt>Bio.Seq</dt><dd><p>Nucleotide sequence with lower efficiency level</p>
</dd>
</dl>
<p>“””</p>
<p>possible_seq_df = self.filtered_sequence(efficiency, self.master_df)</p>
</dd>
</dl>
<blockquote>
<div><p>Exploratory analysis</p>
</div></blockquote>
<p>‘’’</p>
</dd>
</dl>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="contributing.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Contributing</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="installation.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Installation</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Ishaan Joshan Dev &amp; Gayvalin Tammy Sujaritchai
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Usage</a><ul>
<li><a class="reference internal" href="#raises">Raises</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>